### ISSUE TITLE
Is specified version of Python is required when compiling? Compile error using "python setup.py install"

### ISSUE THREAD

**Issue by yuCompHW** [ntucllab/libact#50]
**Hello, Thank you for providing this project**  

**After I have installed the dependencies, I run**
python setup.py install  

**But, I get some errors:**  

Platform Detection: Linux. Link to liblapacke...
running install
running build
running build_py
running build_ext
building 'libact.query_strategies._variance_reduction' extension
C compiler: x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC

compile options: '-I/usr/lib/python2.7/dist-packages/numpy/core/include -I/usr/include/lapacke -I/usr/include/python2.7 -c'
extra options: '-std=c11'
x86_64-linux-gnu-gcc: libact/query_strategies/src/variance_reduction/variance_reduction.c
libact/query_strategies/src/variance_reduction/variance_reduction.c:26:15: error: variable ‘moduledef’ has initializer but incomplete type
 static struct PyModuleDef moduledef = {
               ^
libact/query_strategies/src/variance_reduction/variance_reduction.c:27:5: error: ‘PyModuleDef_HEAD_INIT’ undeclared here (not in a function)
     PyModuleDef_HEAD_INIT,
     ^
  。。。 。。。 。。。
  。。。 。。。 。。。

**I wonder if I need to specify the version of Python, so I tried**
python3 steup.py install
**Still, I cannot install successfully, but the error changes** 
File "setup.py", line 13, in <module>
    from Cython.Build import cythonize
ImportError: No module named 'Cython'  

**However, I have already installed Cython using "pip install Cython"**  

**It will be very kind of you if you could tell me the requirement of version of the installed dependencies**

**OR could you please tell how to modify the "-I/usr/include/lapacke -I/usr/include/python2.7" in the compile option**

**Many Thanks**

**Comment by yangarbiter** [comment]
Currently we supports only python3. Maybe you should try "pip3 install Cython" and see how it works.

**Comment by yuCompHW** [comment]
Thank you very much for your reply, it works for me !

I am now trying to understand the code through running the examples. 

I use **label_digits.py** in the folder example/, and I have some confusions, it will be very kind if you could give me some comments: 

1: in the main function, " _**quota = 10  # ask human to label 30 samples**_"  the value 10 and 30 is confusing, is there a typo?

2: “_**n_classes = 5**_”, I wonder what n_classes refer to? the number of classes of the trained data? but according the getdataset.py, there are only 2 classes of data. I am really confused with this value

3: _**lbr = InteractiveLabeler(label_name=[str(lbl) for lbl in range(n_classes)])**_
As I cannot install matplotlib due to my network, I cannot actually run the function InteractiveLabeler. I wonder if you could tell me a bit more about function InteractiveLabeler:
(a)  is it really retrieve labels from the human input?
(b)  If yes, why this function is called only once? The labeler will  give the labels for all samples ? 
(c)  how many times of interactive labeling are used in the example code? 

Many Thanks

**Comment by yangarbiter** [comment]
Hi yuCompHW,

The documentation for the example label_digits is over here
http://libact.readthedocs.org/en/documents/examples/label_digits.html
We haven't finished it yet, so it is not merged into master.
Sorry for the inconvenience. 

For your questions:

1. Yes, it is a typo. (sorry about that)

2. For this example, we didn't use the dataset from getdataset.py, that one is for plot.py.
The dataset we use for this example is from scikit-learn, it is already inside its package.
No additional download needed.

3. Actually InteractiveLabeler is a class, it constructs a object names ```lbr``` with the name
of each label (the name you would want the human labeler to enter in). In this example, 
whenever you called ```lbr.label(trn_ds.data[ask_id][0].reshape(8, 8))```, it would show
the image of the digit through matplotlib and human labler may enter the corresponding
label through terminal.

Hope this help.
Thank you for reporting your questions.

**Comment by yuCompHW** [comment]
Thank you very much for your kind reply and this is very helpful to me.
I am currently trying to implement an active learning algo for our problem using Scala over Spark. 
I wonder if your team plan to contribute a Spark Mllib library for active learning in the future?
Many Thanks~

**Comment by yuCompHW** [comment]
Hello, sorry to interrupt you~

It will be kind if you could help me solve the following confusions:
In the example/label_digits.py file,
The classifier used in the example is **model = LogisticRegression()**, and according to function **split_train_test** only first **n_labeled** targets in _trn_ds_ are reserved and the others are set to _None_. 
(a) I wonder when using **model.train(trn_ds)**, is the model built only on the samples with **not-None** targets ?
(b) If no, then how can  **sklearn.linear_model.LogisticRegression** perform the semi-supervised classification? According to the scikit-learn's documents, sklearn.linear_model.LogisticRegression is a supervised classifier. 
(c) If yes, is it better to  modify **LogisticRegression** method to a semi-supervised classifier? I wonder if you have passed this consideration?

Many thank for your time and kind help.

**Comment by yangarbiter** [comment]
Currently the models in libact are all supervised learning algorithms. So the model.train simple takes out the labeled data and perform training.

The reason we set the model objects in libact to be supervised learning algorithms are that these models somehow represents the optimization objective in a given active learning problem. In this instance, if you gave LogisticRegression to a query strategy(active learning algorithm), we want the samples being selected are the one that increases the performance of LogisticRegression.

As I know, for now, the active learning algorithms are mostly targeting supervised learning models. Maybe in the future semi-supervised classifiers can be added.

**Comment by yuCompHW** [comment]
Thank you very much for your kind reply. It helps me a lot.

**Comment by yangarbiter** [comment]
No problem!
It seems we may close this issue now.
